pyro/contrib/funsor/handlers/runtime.py

```diff
@@ -0,0 +1,138 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import OrderedDict, namedtuple
+from enum import Enum
+
+
+# name_to_dim : dict, dim_to_name : dict, parents : tuple, iter_parents : tuple
+class StackFrame(namedtuple('StackFrame', [
+            'name_to_dim',
+            'dim_to_name',
+            'parents',
+            'iter_parents',
+            'keep',
+        ])):
+
+    def read(self, name, dim):
+        found_name = self.dim_to_name.get(dim, name)
+        found_dim = self.name_to_dim.get(name, dim)
+        found = name in self.name_to_dim or dim in self.dim_to_name
+        return found_name, found_dim, found
+
+    def write(self, name, dim):
+        assert name is not None and dim is not None
+        self.dim_to_name[dim] = name
+        self.name_to_dim[name] = dim
+
+    def free(self, name, dim):
+        self.dim_to_name.pop(dim, None)
+        self.name_to_dim.pop(name, None)
+        return name, dim
+
+
+class DimType(Enum):
+    """Enumerates the possible types of dimensions to allocate"""
+    LOCAL = 0
+    GLOBAL = 1
+    VISIBLE = 2
+
+
+DimRequest = namedtuple('DimRequest', ['dim', 'dim_type'], defaults=(None, DimType.LOCAL))
+NameRequest = namedtuple('NameRequest', ['name', 'dim_type'], defaults=(None, DimType.LOCAL))
```

Comment:
@eb8680 I think you will need to use the defaults as in [this thread](https://stackoverflow.com/a/18348004).